internal/authentication/oidc/clientcredentials.go

```diff
@@ -0,0 +1,213 @@
+		closeCh:      make(chan struct{}),
+		clock:        clock.RealClock{},
+		fetchTokenFn: conf.Token,
+	}, nil
+}
+
+func (c *ClientCredentials) Run(ctx context.Context) {
+	c.log.Info("Running oidc client_credentials token renewer")
+	renewDuration := c.tokenRenewDuration()
+
+	c.wg.Add(1)
+	go func() {
+		defer func() {
+			c.log.Info("Stopped oidc client_credentials token renewer")
+			c.wg.Done()
+		}()
+
+		for {
+			select {
+			case <-c.closeCh:
+				return
+			case <-ctx.Done():
+				return
+			case <-c.clock.After(renewDuration):
+			}
+
+			c.log.Debug("Renewing client credentials token")
+
+			token, err := c.fetchTokenFn(context.WithValue(ctx, oauth2.HTTPClient, c.httpClient))
+			if err != nil {
+				c.log.Errorf("Error fetching renewed oidc client_credentials token, retrying in 30 seconds: %s", err)
+				renewDuration = time.Second * 30
+				continue
+			}
+
+			c.lock.Lock()
+			c.currentToken = token
+			c.lock.Unlock()
+			renewDuration = c.tokenRenewDuration()
+		}
+	}()
+}
+
+func toConfig(opts ClientCredentialsOptions) (*ccreds.Config, *http.Client, error) {
+	scopes := opts.Scopes
+	if len(scopes) == 0 {
+		// If no scopes are provided, then the default is to use the 'openid' scope
+		// since that is always required for OIDC so implicitly add it.
+		scopes = []string{oidcScopeOpenID}
+	}
+
+	var oidcScopeFound bool
+	for _, scope := range scopes {
+		if scope == oidcScopeOpenID {
+			oidcScopeFound = true
+			break
+		}
+	}
+	if !oidcScopeFound {
+		return nil, nil, fmt.Errorf("oidc client_credentials token source requires the %q scope", oidcScopeOpenID)
+	}
+
+	tokenURL, err := url.Parse(opts.TokenURL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing token URL: %w", err)
+	}
+	if tokenURL.Scheme != "https" {
+		return nil, nil, fmt.Errorf("OIDC token provider URL requires 'https' scheme: %q", tokenURL)
+	}
+
+	conf := &ccreds.Config{
+		ClientID:     opts.ClientID,
+		ClientSecret: opts.ClientSecret,
+		TokenURL:     opts.TokenURL,
+		Scopes:       scopes,
+	}
+
+	if len(opts.Audiences) == 0 {
+		return nil, nil, errors.New("oidc client_credentials token source requires at least one audience")
+	}
+
+	conf.EndpointParams = url.Values{"audience": opts.Audiences}
+
+	// If caPool is nil, then the Go TLS library will use the system's root CA.
+	var caPool *x509.CertPool
+	if len(opts.CAPEM) > 0 {
+		caPool = x509.NewCertPool()
+		if !caPool.AppendCertsFromPEM(opts.CAPEM) {
+			return nil, nil, errors.New("failed to parse CA PEM")
+		}
+	}
+
+	return conf, &http.Client{
+		Timeout: time.Second * 30,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				MinVersion: tls.VersionTLS12,
+				RootCAs:    caPool,
+			},
+		},
+	}, nil
+}
+
+func (c *ClientCredentials) Close() {
+	defer c.wg.Wait()
+
+	if c.closed.CompareAndSwap(false, true) {
+		close(c.closeCh)
+	}
+}
+
+func (c *ClientCredentials) Token() (string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	if c.closed.Load() {
```

Comment:
Here is where you check if the token that's in cache is valid, and if not you lazily request a new one.

Since multiple goroutines could be calling `Token()` at the same time, I recommend implementing some synchronization method (or use the "promise" library!)